src/tamp_improv/benchmarks/base.py

```diff
@@ -0,0 +1,107 @@
+"""Base environment interface."""
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Generic, Protocol, TypeVar
+
+import gymnasium as gym
+from relational_structs import LiftedOperator, PDDLDomain, Predicate, Type
+from task_then_motion_planning.structs import Perceiver, Skill
+
+ObsType = TypeVar("ObsType")
+ActType = TypeVar("ActType")
+
+
+class PredicateContainer(Protocol):
```

Comment:
Yeah totally! It's essentially a more flexible form of inheritance. Unlike traditional inheritance where we have to explicitly inherit (e.g., `class NumberPredicates(PredicateContainer)`), Protocol uses "duck typing" - if a class implements the required methods with the right signatures, it automatically satisfies the protocol.

I choose to use Protocol in this codebase because we have multiple environment implementations that need their own predicate containers, and so I want to get the type checking benefits without forcing a rigid inheritance hierarchy.

But let me know if you think a strict hierarchy makes more sense! I can change it.

Reply:
Cool! Thanks for the explanation. Seems like a good choice here.